File: shared/dtos/llm.py
import re as _re
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, computed_field, field_validator

# Kept in sync with backend/modules/llm/_connections.py::_SLUG_RE. Duplicated
# here to keep the shared package free of backend imports.
_SLUG_RE = _re.compile(r"^[a-z0-9][a-z0-9-]{0,62}$")


class ModelMetaDto(BaseModel):
    connection_id: str
    connection_slug: str = ""  # populated by adapters once Task 2 lands
    connection_display_name: str = ""
    model_id: str
    display_name: str
    context_window: int
    supports_reasoning: bool
    supports_vision: bool
    supports_tool_calls: bool
    parameter_count: str | None = None
    raw_parameter_count: int | None = None
    quantisation_level: str | None = None
    # Upstream has marked this model as deprecated (scheduled for removal).
    # Default False keeps pre-existing cached documents readable — see
    # CLAUDE.md §Data-Model Migrations.
    is_deprecated: bool = False
    # Billing category — "free" (self-hosted / community-shared), "subscription"
    # (covered by a user's upstream plan, e.g. Ollama Cloud or nano-gpt in-plan),
    # or "pay_per_token" (charged per-request, e.g. xAI, Mistral, nano-gpt
    # out-of-plan). Default ``None`` keeps pre-existing cached documents
    # readable — see CLAUDE.md §Data-Model Migrations.
    billing_category: Literal["free", "subscription", "pay_per_token"] | None = None
    # ``True``/``False`` when the upstream provider makes an explicit
    # statement (today: only OpenRouter via ``top_provider.is_moderated``).
    # ``None`` = no statement — every other adapter leaves this default.
    # A future "Allow moderated" filter must handle all three buckets
    # (yes / no / unknown) sensibly. Default ``None`` keeps pre-existing
    # cached documents readable — see CLAUDE.md §Data-Model Migrations.
    is_moderated: bool | None = None

    @computed_field
    @property
    def unique_id(self) -> str:
        return f"{self.connection_slug}:{self.model_id}"


class UserModelConfigDto(BaseModel):
    model_unique_id: str
    is_favourite: bool = False
    is_hidden: bool = False
    custom_display_name: str | None = None
    custom_context_window: int | None = None
    # Override for upstream-reported reasoning capability. Primary use case:
    # community/homelab sidecars that don't yet detect every thinker family.
    # ``None`` = respect upstream. ``True``/``False`` = force the flag.
    custom_supports_reasoning: bool | None = None
    notes: str | None = None
    system_prompt_addition: str | None = None


class SetUserModelConfigDto(BaseModel):
    is_favourite: bool | None = None
    is_hidden: bool | None = None
    custom_display_name: str | None = None
    custom_context_window: int | None = None
    custom_supports_reasoning: bool | None = None
    notes: str | None = None
    system_prompt_addition: str | None = None

    @field_validator("custom_display_name")
    @classmethod
    def validate_display_name(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if len(v) == 0:
            return None
        if len(v) > 100:
            raise ValueError("custom_display_name must be 100 characters or fewer")
        return v

    @field_validator("custom_context_window")
    @classmethod
    def validate_context_window(cls, v: int | None) -> int | None:
        if v is None:
            return None
        if v < 80_000:
            raise ValueError("custom_context_window must be at least 80000")
        return v


class AdapterTemplateDto(BaseModel):
    id: str
    display_name: str
    slug_prefix: str
    config_defaults: dict[str, Any]
    required_config_fields: list[str] = Field(default_factory=list)


class AdapterDto(BaseModel):
    adapter_type: str
    display_name: str
    view_id: str
    templates: list[AdapterTemplateDto]
    config_schema: list[dict[str, Any]]
    secret_fields: list[str]


class ConnectionDto(BaseModel):
    id: str
    user_id: str
    adapter_type: str
    display_name: str
    slug: str
    config: dict[str, Any]
    last_test_status: Literal["untested", "valid", "failed"] | None = None
    last_test_error: str | None = None
    last_test_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    # Connections managed by the system (e.g. a host-self connection that mirrors
    # a Homelab) cannot be edited or deleted via the generic connection endpoints.
    is_system_managed: bool = False


class CreateConnectionDto(BaseModel):
    adapter_type: str
    display_name: str
    slug: str
    config: dict[str, Any]


class UpdateConnectionDto(BaseModel):
    display_name: str | None = None
    slug: str | None = None
    config: dict[str, Any] | None = None


# --- Community Provisioning ---


class HomelabEngineInfoDto(BaseModel):
    type: str
    version: str | None = None


class HomelabDto(BaseModel):
    homelab_id: str
    display_name: str
    host_key_hint: str
    status: Literal["active", "revoked"]
    created_at: datetime
    last_seen_at: datetime | None = None
    last_sidecar_version: str | None = None
    last_engine_info: HomelabEngineInfoDto | None = None
    is_online: bool = False
    # Process-local concurrency cap shared across every consumer of this
    # homelab (host-self + every api-key-backed Connection). Default 3 keeps
    # pre-existing documents readable (see CLAUDE.md §Data-Model Migrations).
    max_concurrent_requests: int = 3
    # Slug under which a system-managed self-Connection is registered for the
    # host. ``None`` for homelabs created before this field existed; such
    # homelabs don't have a self-Connection — the host continues to use api-keys.
    host_slug: str | None = None


class HomelabCreatedDto(HomelabDto):
    plaintext_host_key: str = Field(
        ..., description="Shown exactly once; never returned again."
    )


class HomelabHostKeyRegeneratedDto(HomelabDto):
    plaintext_host_key: str = Field(
        ..., description="Shown exactly once; never returned again."
    )


class HomelabStatusDto(BaseModel):
    """Lightweight connection-status snapshot.

    Emitted alongside ``llm.homelab.status_changed`` / ``llm.homelab.last_seen``
    when Plan 3 (CSP sidecar) updates a homelab's live state without needing
    to re-send the full :class:`HomelabDto`.
    """

    homelab_id: str
    is_online: bool
    last_seen_at: datetime | None = None
    last_sidecar_version: str | None = None
    last_engine_info: HomelabEngineInfoDto | None = None


class CreateHomelabDto(BaseModel):
    display_name: str
    host_slug: str
    max_concurrent_requests: int = 3

    @field_validator("display_name")
    @classmethod
    def _name_len(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("display_name must not be empty")
        if len(v) > 80:
            raise ValueError("display_name must be 80 characters or fewer")
        return v

    @field_validator("host_slug")
    @classmethod
    def _slug_shape(cls, v: str) -> str:
        v = v.strip().lower()
        if not _SLUG_RE.match(v):
            raise ValueError(
                "host_slug must be 1-63 chars, lowercase alphanumeric with hyphens, "
                "starting with a letter or digit"
            )
        return v

    @field_validator("max_concurrent_requests")
    @classmethod
    def _mcr_bounds(cls, v: int) -> int:
        if v < 1 or v > 64:
            raise ValueError("max_concurrent_requests must be between 1 and 64")
        return v


class UpdateHomelabDto(BaseModel):
    display_name: str | None = None
    max_concurrent_requests: int | None = None

    @field_validator("display_name")
    @classmethod
    def _name_len(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if not v:
            raise ValueError("display_name must not be empty")
        if len(v) > 80:
            raise ValueError("display_name must be 80 characters or fewer")
        return v

    @field_validator("max_concurrent_requests")
    @classmethod
    def _mcr_bounds(cls, v: int | None) -> int | None:
        if v is None:
            return None
        if v < 1 or v > 64:
            raise ValueError("max_concurrent_requests must be between 1 and 64")
        return v


class ApiKeyDto(BaseModel):
    api_key_id: str
    homelab_id: str
    display_name: str
    api_key_hint: str
    allowed_model_slugs: list[str]
    status: Literal["active", "revoked"]
    created_at: datetime
    revoked_at: datetime | None = None
    last_used_at: datetime | None = None
    # Per-api-key concurrency cap enforced in CommunityAdapter. Default 1 keeps
    # pre-existing documents readable (see CLAUDE.md §Data-Model Migrations).
    max_concurrent: int = 1


class ApiKeyCreatedDto(ApiKeyDto):
    plaintext_api_key: str = Field(
        ..., description="Shown exactly once; never returned again."
    )


class CreateApiKeyDto(BaseModel):
    display_name: str
    allowed_model_slugs: list[str] = Field(default_factory=list)
    max_concurrent: int = 1

    @field_validator("display_name")
    @classmethod
    def _name_len(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("display_name must not be empty")
        if len(v) > 80:
            raise ValueError("display_name must be 80 characters or fewer")
        return v

    @field_validator("max_concurrent")
    @classmethod
    def _mc_bounds(cls, v: int) -> int:
        if v < 1 or v > 32:
            raise ValueError("max_concurrent must be between 1 and 32")
        return v


class UpdateApiKeyDto(BaseModel):
    display_name: str | None = None
    allowed_model_slugs: list[str] | None = None
    max_concurrent: int | None = None

    @field_validator("display_name")
    @classmethod
    def _name_len(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if not v:
            raise ValueError("display_name must not be empty")
        if len(v) > 80:
            raise ValueError("display_name must be 80 characters or fewer")
        return v

    @field_validator("max_concurrent")
    @classmethod
    def _mc_bounds(cls, v: int | None) -> int | None:
        if v is None:
            return None
        if v < 1 or v > 32:
            raise ValueError("max_concurrent must be between 1 and 32")
        return v

File: frontend/src/core/types/llm.ts
export type AdapterConfigFieldType = 'string' | 'url' | 'secret' | 'integer'

export interface AdapterConfigFieldHint {
  name: string
  type: AdapterConfigFieldType
  label: string
  required: boolean
  min: number | null
  max: number | null
  placeholder: string | null
}

export interface AdapterTemplate {
  id: string
  display_name: string
  slug_prefix: string
  config_defaults: Record<string, unknown>
  required_config_fields: string[]
}

export interface Adapter {
  adapter_type: string
  display_name: string
  view_id: string
  templates: AdapterTemplate[]
  config_schema: AdapterConfigFieldHint[]
  secret_fields: string[]
}

/**
 * Secret fields are redacted in the server payload:
 * `config[secret_field] = { is_set: boolean }`. Plain fields pass through.
 */
export type SecretFieldView = { is_set: boolean }

export interface Connection {
  id: string
  user_id: string
  adapter_type: string
  display_name: string
  slug: string
  config: Record<string, unknown>
  last_test_status: 'untested' | 'valid' | 'failed' | null
  last_test_error: string | null
  last_test_at: string | null   // ISO string on the wire
  created_at: string
  updated_at: string
  /**
   * System-managed connections are created and maintained by other backend
   * subsystems (e.g. the host's own homelab self-connection). They cannot
   * be edited or deleted via the generic Connection API — mutate the owning
   * resource instead (e.g. the Homelab).
   * Optional for backwards compat on older payloads — treat missing as false.
   */
  is_system_managed?: boolean
}

export interface CreateConnectionRequest {
  adapter_type: string
  display_name: string
  slug: string
  config: Record<string, unknown>
}

export interface UpdateConnectionRequest {
  display_name?: string
  slug?: string
  config?: Record<string, unknown>
}

/**
 * Discrete reasoning-effort buckets exposed by a model (e.g. "low", "medium",
 * "high"). Mirrors ``shared/dtos/llm.py::ReasoningEffortSpec``.
 */
export interface ReasoningEffortSpec {
  buckets: string[]
  default_bucket: string
}

/**
 * How a model supports reasoning. ``no_reasoning`` = the model cannot reason;
 * ``optional`` = caller decides per-request; ``always_on`` = the model always
 * reasons (no off switch).
 */
export interface ReasoningCapability {
  kind: 'no_reasoning' | 'optional' | 'always_on'
  effort: ReasoningEffortSpec | null
  default_on: boolean
}

/**
 * Tool-calling capability of a model. ``exclusive_with_reasoning`` = the
 * provider rejects requests that combine tools and reasoning, so the cockpit
 * must keep them mutually exclusive.
 */
export interface ToolCapability {
  supported: boolean
  exclusive_with_reasoning: boolean
}

/**
 * Resolved per-model capabilities — what the cockpit needs to know to render
 * the right controls. ``first_class_support`` is set by the adapter when it
 * has explicit knowledge of the model (vs a best-effort fallback).
 */
export interface ResolvedCapabilities {
  reasoning: ReasoningCapability
  tools: ToolCapability
  first_class_support: boolean
}

export interface ModelMetaDto {
  connection_id: string
  connection_slug: string
  connection_display_name: string
  model_id: string
  display_name: string
  context_window: number
  /**
   * Computed on the backend from ``reasoning.kind != "no_reasoning"``. Kept
   * here for legacy consumers that haven't migrated to the structured
   * ``reasoning`` capability yet.
   */
  supports_reasoning: boolean
  supports_vision: boolean
  supports_tool_calls: boolean
  /** Structured reasoning capability — preferred over ``supports_reasoning``. */
  reasoning: ReasoningCapability
  /** Structured tool capability — preferred over ``supports_tool_calls``. */
  tools: ToolCapability
  /**
   * True when the adapter has first-class knowledge of this model (curated
   * capability rules). False = best-effort defaults; the cockpit may want
   * to dim the indicator or surface a "limited info" hint.
   */
  first_class_support: boolean
  parameter_count: string | null
  raw_parameter_count: number | null
  quantisation_level: string | null
  /**
   * Upstream has marked this model as deprecated (scheduled for removal).
   * Optional for backwards compat on older cached payloads — treat missing
   * as false. See shared/dtos/llm.py::ModelMetaDto.is_deprecated.
   */
  is_deprecated?: boolean
  /**
   * How this model bills the user. ``free`` = no cost, ``subscription`` =
   * covered by an upstream plan (Ollama Cloud, nano-gpt subscription tier),
   * ``pay_per_token`` = metered. Optional for backwards compat with older
   * cached payloads — treat missing/null as "unknown".
   */
  billing_category?: 'free' | 'subscription' | 'pay_per_token' | null
  /**
   * Free-form per-model note rendered as a dimmed third line under the
   * row in the model browser. Adapter-set on the backend (`shared/dtos/
   * llm.py::ModelMetaDto.remarks`); the frontend treats it as read-only.
   * Optional/null for every model that has nothing to disclose.
   */
  remarks?: string | null
  /**
   * True when the model is hosted under a privacy-preserving guarantee
   * (e.g. confidential compute, no logging). Adapter-set on the backend
   * (`shared/dtos/llm.py::ModelMetaDto.is_privacy_preserving`). Optional
   * for backwards compat with older cached payloads — treat missing as
   * false.
   */
  is_privacy_preserving?: boolean
  unique_id: string
}

/** Model enriched with the user's per-model configuration (merged client-side). */
export interface EnrichedModelDto extends ModelMetaDto {
  user_config: UserModelConfigDto | null
}

export interface UserModelConfigDto {
  model_unique_id: string
  is_favourite: boolean
  is_hidden: boolean
  custom_display_name: string | null
  custom_context_window: number | null
  /** null = honour upstream. true/false = force the flag. */
  custom_supports_reasoning: boolean | null
  notes: string | null
  system_prompt_addition: string | null
}

export interface SetUserModelConfigRequest {
  is_favourite?: boolean
  is_hidden?: boolean
  custom_display_name?: string | null
  custom_context_window?: number | null
  custom_supports_reasoning?: boolean | null
  notes?: string | null
  system_prompt_addition?: string | null
}

export interface TestResultResponse {
  valid: boolean
  error: string | null
}
